### PR TITLE
fix(remap): prevent lexer getting stuck on an unterminated literal

### LINKED ISSUE #6875: VRL REPL hangs for `parse_grok` input
Reported by user in discord: https://discord.com/channels/742820443487993987/746070591097798688/824146032543399967

The following input in the VRL console hangs. I think it is waiting for more input:

```
parse_grok!("1.2.3.4 - - [23/Mar/2021:06:46:35 +0000]", "%{IPORHOST:remote_ip} %{USER:ident} %{USER:user_name} \[%{HTTPDATE:timestamp}\]")
```

Possibly related to https://github.com/timberio/vector/issues/6788

### LINKED ISSUE #6879: High CPU / freeze, when regex is missing the closing quote
### Vector Version

```
version="0.12.1" git_version="v0.12.1" released="Fri, 12 Mar 2021 21:27:28 +0000" arch="x86_64"
```

### Vector Configuration File

```toml
[sources.ftth_syslog]
  type = "socket"
  mode = "udp"
  address = "0.0.0.0:5140"

[transforms.parse_syslog]
  type = "remap"
  inputs = ["ftth_syslog"]
  source = '''
  . = parse_regex!(.message, r'(?P<syslog_pri><\d+>)(?P<syslog_seqnum>\d+)(?:\W+)(?P<syslog_host>\w+-\w+-\w+)(?:\W+)(?P<syslog_datetime>.*?)(?:\W+)(?P<message_type>\.*?)(?:\W+)(?P<syslog_message>.*))
  '''

[sinks.console_out]
  inputs = ["parse_syslog"]
  type = "console"
  encoding = "json"
```

### Debug Output

```
Mar 25 09:54:08.589 DEBUG sink{component_kind="sink" component_name=console_out component_type=console}: vector::topology::builder: Finished.
Mar 25 09:54:09.095  INFO vector::app: Log level is enabled. level="debug"
Mar 25 09:54:09.096  INFO vector::sources::host_metrics: PROCFS_ROOT is unset. Using default '/proc' for procfs root.
Mar 25 09:54:09.096  INFO vector::sources::host_metrics: SYSFS_ROOT is unset. Using default '/sys' for sysfs root.
Mar 25 09:54:09.096  INFO vector::app: Loading configs. path=[("/etc/vector/vector.toml", Some(TOML))]
```

### Expected Behavior

Vector should catch the error and fail to start.

### Actual Behavior

Vector did start, printed out the above logs and took over the CPU (100% load).

### Example Data

Example of Syslog messages:
```
<180>344886: EAC1106-Hveragerdi-A: Mar 25 09:53:35: %IP_SOURCE_GUARD-4-DENY_INVALID_PACKET: Detected and dropped illegal traffic on port Gi5/32 and vlan 106, the non-cumulative packet dropped count is 2.
<180>344885: EAC1106-Hveragerdi-A: Mar 25 09:53:32: %SW_DAI-4-INVALID_ARP: 3 Invalid ARPs (Req) on Gi6/76, vlan 999.([5c85.7e43.3aaa/0.0.0.0/0000.0000.0000/169.254.191.78/09:52:32 UTC Thu Mar 25 2021])
```

### PR DESCRIPTION
**Pull request by FungusHumungus**
Closes #6879 
Closes #6875 

The lexer was getting stuck in an infinite loop when it reached an unclosed quote. This changes the `query_start` function so that it returns a `Result` which allows the function to return an error in that situation.

Turns out the lexer was getting stuck whenever it was lexing a literal from within function arguments as it was ignoring any errors raised from lexing the literals and got stuck lexing the same text repeatedly.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: lib/vrl/parser/src/lex.rs [hunk 1]
```diff
@@ -62,6 +65,7 @@ impl DiagnosticError for Error {
             Literal { .. } => 208,
             EscapeChar { .. } => 209,
             UnexpectedParseError(..) => 210,
+            UnterminatedLiteral { .. } => 211,
```
Comment (jszwedko) [inline lib/vrl/parser/src/lex.rs:68]: Should we add this to the VRL error docs?

**Review by jszwedko — COMMENTED** [review]
Nice! This seems to work.

It actually causes the case in #6875 to no longer hang too, but it shows the unterminated literal error which I don't think is right for that input of:

```
parse_grok!("1.2.3.4 - - [23/Mar/2021:06:46:35 +0000]", "%{IPORHOST:remote_ip} %{USER:ident} %{USER:user_name} \[%{HTTPDATE:timestamp}\]")
```

Here, Jean mentioned (https://github.com/timberio/vector/issues/6875#issuecomment-808048403) that the issue is that `\[` is not a valid escape sequence. Is it possible that the error handling for `literal_check` that you have is overly broad and that we need to differentiate the returned errors?

File: lib/vrl/parser/src/lex.rs (hunk 1 above)
Comment (FungusHumungus) [reply lib/vrl/parser/src/lex.rs:68]: > It actually causes the case in #6875 to no longer hang too

That's a surprise! I'll look into that.

It doesn't seem that that error number surfaces in the UI, a 202 parse error gets reported.

**Comment by FungusHumungus** [comment]
@jszwedko Good spot! It turns out it was getting stuck when lexing any errors in a literal and not just unterminated literals. I have changed things now so that it bubbles up the error from lexing the literal rather than creating a new error so the issues you noted should be ok now.

**Comment by jszwedko** [comment]
This looks good, thanks @FungusHumungus !

File: lib/vrl/parser/src/lex.rs [hunk 2]
```diff
@@ -751,49 +755,70 @@ impl<'input> Lexer<'input> {
...
 
                         let ch = match &self.input[pos..] {
                             s if s.starts_with('"') => {
-                                let r = Lexer::new(&self.input[pos + 1..]).string_literal(0);
+                                let r = Lexer::new(&self.input[pos + 1..]).string_literal(0)?;
                                 match literal_check(r, &mut chars) {
                                     Ok(ch) => ch,
-                                    Err(_) => continue,
+                                    Err(_) => {
+                                        // The call to lexer about should have raised an appropriate error by now,
```
Comment (JeanMertz) [inline lib/vrl/parser/src/lex.rs:779]:
```suggestion
                                        // The call to lexer above should have raised an appropriate error by now,
```

**Review by JeanMertz — APPROVED** [review]

**Pull request merged**